core/src/main/java/org/apache/spark/shuffle/sort/BypassMergeSortShuffleWriter.java

```diff
@@ -216,6 +228,60 @@ public void write(Iterator<Product2<K, V>> records) throws IOException {
     return lengths;
   }
 
+  private long[] combineAndWritePartitionsUsingPluggableWriter() throws IOException {
+    // Track location of the partition starts in the output file
+    final long[] lengths = new long[numPartitions];
+    if (partitionWriters == null) {
+      // We were passed an empty iterator
+      return lengths;
+    }
+    assert(pluggableWriteSupport != null);
+
+    final long writeStartTime = System.nanoTime();
+    ShuffleMapOutputWriter mapOutputWriter = pluggableWriteSupport.newMapOutputWriter(
+          appId, shuffleId, mapId);
+    try {
+      for (int i = 0; i < numPartitions; i++) {
+        final File file = partitionWriterSegments[i].file();
+        if (file.exists()) {
+          final FileInputStream in = new FileInputStream(file);
+          boolean copyThrewException = true;
+          ShufflePartitionWriter writer = mapOutputWriter.newPartitionWriter(i);
+          try {
+            try (OutputStream out = writer.openPartitionStream()) {
+              Utils.copyStream(in, out, false, false);
```

Comment:
any reason you're not supporting `transferTo` here?  in fact, I wonder if your api should actually expose `java.nio.channels.WritableByteChannel` as the destination, to allow for a zero-copy transfer.  Even if you're writing to a remote destination, that would let the OS to move data from the file to the socket without copying into user space, and back out  (aka zero-copy).  I'm not sure exactly how this would effect the api, though, you might want to let ShufflePartitionWriters choose to expose either OutputStream or WritableByteChannel

Reply:
Mostly because for a first prototype I wanted to keep things simple, and I think there's a chance that exposing channel level methods in the API would be confusing for devs. This stuff is already pretty low level as it is, and adding an additional switch in the API to support another low level option is only going to multiply the complexity even further.

I also don't fully understand how `transferTo` works in the first place in the context where we're not writing files to local disk. A suggestion on how we could support that as a patch on top of this work would be helpful to get us started.

For what it's worth I don't see many non-Spark native implementations being able to support channel based writes. Internally I'm experimenting with another storage system that certainly won't be able to support channel based write. If we did include channel support, it should be strictly optional while outputStream based write is mandatory. So I see channel support as being additive to this work and not replacing it.